### ISSUE TITLE
OSX: ld: can't link with bundle (MH_BUNDLE) only dylibs (MH_DYLIB) file '../_build/test/../libxmlm.so' for architecture x86_64

### ISSUE THREAD

**Issue by djs55** [yallop/ocaml-ctypes-inverted-stubs-example#4]
```
Davids-MacBook-Pro:test djs$ make
gcc -c -o ../_build/test/test.o -I /usr/local/lib/ocaml -I ../_build/test/../generated  test.c
gcc -o ../_build/test/test.native ../_build/test/test.o -L../_build/test/.. -lxmlm
ld: can't link with bundle (MH_BUNDLE) only dylibs (MH_DYLIB) file '../_build/test/../libxmlm.so' for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [../_build/test/test.native] Error 1
```

I notice that the .so file is one of these:

```
../_build/libxmlm.so: Mach-O 64-bit bundle x86_64
```

whereas a working .so I have from another project is one of these:

```
Mach-O 64-bit dynamically linked shared library x86_64
```

For reference the library built like this:

```
ocamlfind opt -c -o _build/lib/bindings.cmx -I _build/generated -I _build/lib -package xmlm,ctypes.stubs,ctypes.foreign lib/bindings.ml
ocamlfind opt -c -o _build/stub_generator/generate.cmx -I _build/generated -I _build/lib -package xmlm,ctypes.stubs,ctypes.foreign stub_generator/generate.ml
ocamlfind opt -o _build/generate -linkpkg -package xmlm,ctypes.stubs,ctypes.foreign _build/lib/bindings.cmx _build/stub_generator/generate.cmx
_build/generate _build/generated
ocamlfind opt -c -o _build/generated/xmlm_bindings.cmx -I _build/generated -I _build/lib -package xmlm,ctypes.stubs,ctypes.foreign _build/generated/xmlm_bindings.ml
ocamlfind opt -c -o _build/lib/apply_bindings.cmx -I _build/generated -I _build/lib -package xmlm,ctypes.stubs,ctypes.foreign lib/apply_bindings.ml
gcc -c -o _build/generated/xmlm.o -fPIC -I /Users/djs/.opam/system/lib/ctypes -I /usr/local/lib/ocaml -I /usr/local/lib/ocaml/../ctypes _build/generated/xmlm.c
ocamlfind opt -o _build/libxmlm.so -linkpkg -output-obj -runtime-variant _pic -verbose -package xmlm,ctypes.stubs,ctypes.foreign _build/lib/bindings.cmx _build/generated/xmlm_bindings.cmx _build/lib/apply_bindings.cmx _build/generated/xmlm.o
Effective set of compiler predicates: pkg_xmlm,pkg_unix,pkg_bigarray,pkg_str,pkg_bytes,pkg_ctypes,pkg_ctypes.stubs,pkg_ctypes.foreign.base,pkg_ctypes.foreign.unthreaded,pkg_ctypes.foreign,autolink,native
+ ocamlopt.opt -o _build/libxmlm.so -output-obj -runtime-variant _pic -verbose -I /Users/djs/.opam/system/lib/xmlm -I /Users/djs/.opam/system/lib/bytes -I /Users/djs/.opam/system/lib/ctypes /Users/djs/.opam/system/lib/xmlm/xmlm.cmxa /usr/local/lib/ocaml/unix.cmxa /usr/local/lib/ocaml/bigarray.cmxa /usr/local/lib/ocaml/str.cmxa /Users/djs/.opam/system/lib/ctypes/ctypes.cmxa /Users/djs/.opam/system/lib/ctypes/cstubs.cmxa /Users/djs/.opam/system/lib/ctypes/ctypes-foreign-base.cmxa /Users/djs/.opam/system/lib/ctypes/ctypes-foreign-unthreaded.cmxa _build/lib/bindings.cmx _build/generated/xmlm_bindings.cmx _build/lib/apply_bindings.cmx _build/generated/xmlm.o
+ clang -arch x86_64 -c -o '/var/folders/ml/fqqlxl6j03l2lf9ljfms9m540000gn/T/camlstartupc91edd.o' '/var/folders/ml/fqqlxl6j03l2lf9ljfms9m540000gn/T/camlstartup92c2d5.s'
+ gcc -bundle -flat_namespace -undefined suppress -Wl,-no_compact_unwind -o '_build/libxmlm.so'   '-L/Users/djs/.opam/system/lib/xmlm' '-L/Users/djs/.opam/system/lib/bytes' '-L/Users/djs/.opam/system/lib/ctypes' '-L/usr/local/lib/ocaml'  '/var/folders/ml/fqqlxl6j03l2lf9ljfms9m540000gn/T/camlstartupc91edd.o' '_build/lib/apply_bindings.o' '_build/generated/xmlm_bindings.o' '_build/lib/bindings.o' '/Users/djs/.opam/system/lib/ctypes/ctypes-foreign-unthreaded.a' '/Users/djs/.opam/system/lib/ctypes/ctypes-foreign-base.a' '/Users/djs/.opam/system/lib/ctypes/cstubs.a' '/Users/djs/.opam/system/lib/ctypes/ctypes.a' '/usr/local/lib/ocaml/str.a' '/usr/local/lib/ocaml/bigarray.a' '/usr/local/lib/ocaml/unix.a' '/Users/djs/.opam/system/lib/xmlm/xmlm.a' '/usr/local/lib/ocaml/stdlib.a' '-L/usr/local/Cellar/libffi/3.0.13/lib' '-lffi' '-lctypes-foreign-base_stubs' '-L/usr/local/Cellar/libffi/3.0.13/lib' '-lffi' '-lctypes_stubs' '-lcamlstr' '-lbigarray' '-lunix' '_build/generated/xmlm.o' '/usr/local/lib/ocaml/libasmrun_pic.a' 
```

I suspect the problem is the use of `-bundle` instead of `-shared`. I added `-ccopt -shared` but this fails with:

```
+ gcc -bundle -flat_namespace -undefined suppress -Wl,-no_compact_unwind -o '_build/libxmlm.so'   '-L/Users/djs/.opam/system/lib/xmlm' '-L/Users/djs/.opam/system/lib/bytes' '-L/Users/djs/.opam/system/lib/ctypes' '-L/usr/local/lib/ocaml' -shared '/var/folders/ml/fqqlxl6j03l2lf9ljfms9m540000gn/T/camlstartup3d473f.o' '_build/lib/apply_bindings.o' '_build/generated/xmlm_bindings.o' '_build/lib/bindings.o' '/Users/djs/.opam/system/lib/ctypes/ctypes-foreign-unthreaded.a' '/Users/djs/.opam/system/lib/ctypes/ctypes-foreign-base.a' '/Users/djs/.opam/system/lib/ctypes/cstubs.a' '/Users/djs/.opam/system/lib/ctypes/ctypes.a' '/usr/local/lib/ocaml/str.a' '/usr/local/lib/ocaml/bigarray.a' '/usr/local/lib/ocaml/unix.a' '/Users/djs/.opam/system/lib/xmlm/xmlm.a' '/usr/local/lib/ocaml/stdlib.a' '-L/usr/local/Cellar/libffi/3.0.13/lib' '-lffi' '-lctypes-foreign-base_stubs' '-L/usr/local/Cellar/libffi/3.0.13/lib' '-lffi' '-lctypes_stubs' '-lcamlstr' '-lbigarray' '-lunix' '_build/generated/xmlm.o' '/usr/local/lib/ocaml/libasmrun_pic.a' 
clang: error: invalid argument '-bundle' not allowed with '-dynamiclib'
File "caml_startup", line 1:
Error: Error during linking
ocamlopt.opt returned with exit code 2
```

I tried `-shared` but this seems to conflict with `-output-obj`. Removing `-output-obj` didn't seem to fix the problem either.

**Comment by yallop** [comment]
Reproduced here: https://travis-ci.org/yallop/ocaml-ctypes-inverted-stubs-example/jobs/90404440#L708
Fixes welcome!

**Comment by abbysmal** [comment]
Hi,

We encountered the same error with @dinosaure.
We fixed it by removing output-obj and adding -dynamiclib.
Maybe the problem is that output-obj somehow use -bundle instead of -dynamiclib, so it can be fixed by justing specifying the right flag instead of relying on -output-obj.
Is there a reason why -output-obj use -bundle instead of -dynamiclib ?

Cheers,

**Comment by yallop** [comment]
> We fixed it by removing output-obj and adding -dynamiclib.

That sounds promising.  Could you propose a patch to the Makefile in this repository?

> Is there a reason why -output-obj use -bundle instead of -dynamiclib ?

I don't know, but it would be good to raise the issue on [Mantis](http://caml.inria.fr/mantis/) if `-bundle` is the wrong choice.

**Comment by yallop** [comment]
@Engil has fixed this in #12.

**Comment by msprotz** [comment]
Just hit this. Has anyone raised a corresponding issue on Mantis? It seems strange that -bundle should be the default?
